### PR TITLE
[SQL] Support INTERSECT/EXCEPT ALL by rewriting using the new `positive` operator

### LINKED ISSUE #5483: [DBSP][SQL] `EXCEPT ALL` is not implemented
This could be implemented by adapting the DBSP operator DistinctIncrementalTotal.

### LINKED ISSUE #6436: [SQL] INSERSECT ALL is not supported

### PR DESCRIPTION
**Pull request by mihaibudiu**
Fixes #6436
Fixes #5483 

`EXCEPT ALL` is implemented using the newly-merged DBSP `positive` operator.
`A INTERSECT ALL B` is rewritten as `A EXCEPT ALL (A EXCEPT ALL B)`. A similar rule is used for simple `INTERSECT` instead of synthesizing a join.

### Describe Manual Test Plan

Ran all the new Java tests.

## Checklist

- [x] Unit tests added/updated
- [x] Integration tests added/updated
- [x] Documentation updated

### CODE REVIEW

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java [hunk 1]
```diff
@@ -46,7 +46,7 @@ public void issue3418() {
     @Test
     public void testInternal179() {
         var cc = this.getCC("""
-                CREATE TABLE h(p REAL NOT NULL);
+                CREATE TABLE h(p REAL NOT NULL);S
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/RegressionTests.java]: Stray `S` after the semicolon — this will cause a SQL parse error in the test. Looks like a typo introduced during the `step` → `stepWeightOne` refactor.

File: docs.feldera.com/docs/sql/unsupported-operations.md [hunk 2]
```diff
@@ -100,8 +100,9 @@ supported.
 columns. Refer to [PIVOT documentation](./aggregates.md#pivots) for
 example usage.  Dynamic `PIVOT` is not yet supported.
 
-## `INTERSECT ALL` and `EXCEPT ALL`
-`INTERSECT ALL` and `EXCEPT ALL` operations are not yet supported.
+## `EXCEPT ALL`
+
+`EXCEPT ALL` operations is not yet supported.
```
Comment (mythical-fred) [inline docs.feldera.com/docs/sql/unsupported-operations.md]: Nit: "operations is" → "operation is" (subject-verb agreement).

**Review by mythical-fred — CHANGES_REQUESTED** [review]
Two issues — one real bug, one prose nit. The rewrite rule itself is clean and well-documented; good test coverage including NULL semantics and multi-way INTERSECT ALL.

Also: commit subject says "INSERSECT" — worth fixing with `git rebase -i` before merge.

**Review by mythical-fred — APPROVED** [review]
LGTM

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/SetOpTests.java [hunk 3]
```diff
@@ -0,0 +1,575 @@
...
+                 1
+                (2 rows)""");
+    }
+
+    @Test
+    public void testExceptEmpty() {
+        // Right side contains all left values: result is empty.
+        this.qst("""
+                SELECT * FROM (VALUES (1), (2)) AS T(x)
+                EXCEPT
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/SetOpTests.java:594]: Non-blocking: please add `testExceptAll*` tests symmetric to the `testIntersectAll*` set: empty inputs, multi-column, three-way (left-associative), and the `count_B(v) >= count_A(v)` boundary cases. Currently EXCEPT ALL is covered only by the two inline cases in `calciteExcept`, which is much thinner than the INTERSECT ALL coverage in this same file.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/optimizer/ExceptAllFoldRule.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        List<RelNode> inputs = minus.getInputs();
+        assert inputs.size() > 2;
+
+        RelBuilder builder = call.builder();
+
+        // Push left side (input[0]).
+        builder.push(inputs.get(0));
+
+        // UNION ALL inputs[1..n-1] into a single right-side relation.
+        builder.push(inputs.get(1));
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/optimizer/ExceptAllFoldRule.java:50]: FYI: Calcite's SQL parser emits `A EXCEPT ALL B EXCEPT ALL C` as nested binary `Minus` nodes, so this rule only fires if something upstream (e.g. `MinusMergeRule`) merges them into an n-ary node. I don't see `MinusMergeRule` registered in the optimizer pipeline, so this rule may be dead code today. Harmless either way — leaving it in is reasonable defensive coding — but worth a one-line comment about the precondition, or a test that exercises an n-ary `Minus` to make sure the path is alive.

**Review by mythical-fred — APPROVED** [review]
Re-review for EXCEPT ALL scope expansion (prior approval was at 26fa6261d0 for INTERSECT ALL only).

## Verdict: LGTM

The EXCEPT ALL implementation mirrors the (already-reviewed) INTERSECT ALL one and is correct.

### Correctness notes
- `ExceptAllToJoinRule`: LEFT JOIN A vs B on `IS NOT DISTINCT FROM` over data cols + `$rn`, filtered by `b.$rn IS NULL`. A row `(v, k)` from A survives iff B has no row `(v, k)`, i.e. iff `k > count_B(v)`. Surviving k's are `count_B(v)+1 .. count_A(v)`, giving exactly `max(0, count_A(v) - count_B(v))` copies. Multiset-difference semantics. ✓
- Field-offset arithmetic checks out: left `$rn` at position `fieldCount`, right `$rn` at `2*fieldCount + 1`. ✓
- Nice reuse of `IntersectAllToJoinRule.augmentWithRowNumber` from `ExceptAllToJoinRule`. ✓
- `ExceptAllFoldRule` (n>2 → binary) is defensive: only fires if Calcite produced an n-ary `Minus` upstream (e.g. via `MinusMergeRule`). Harmless if it never matches; the parser-emitted left-associative nested binaries are handled by `ExceptAllToJoinRule` directly.
- `CalciteToDBSPCompiler.visitMinus/visitIntersect` correctly converted from `UnimplementedException` to `InternalCompilerError` — good fail-loud if the rule didn't fire.

### Test coverage gap (non-blocking, please consider)
`SetOpTests.java` has a thorough `testIntersectAll*` series (basic, minCounts, empty, null, nullBothSides, multiColumn, threeWay) but **no symmetric `testExceptAll*` series**. EXCEPT ALL is only covered by two inline cases in `calciteExcept` (basic + nulls). Worth adding tests for:
- empty left / empty right
- multi-column rows
- three-way left-associative `A EXCEPT ALL B EXCEPT ALL C`
- `count_B(v) > count_A(v)` edge (result has 0 copies)
- `count_B(v) == count_A(v)` edge (all consumed)

Not a blocker — `calciteExcept` plus the planner symmetry with INTERSECT ALL give reasonable confidence.

### Minor
- `IS_NOT_DISTINCT_FROM` on `$rn` is fine (ROW_NUMBER is non-null so it reduces to `=`); a plain `=` on `$rn` would be slightly clearer but won't change planning.

Approved.

**Comment by mihaibudiu** [comment]
Needs #6472 to be merged first

**Review by mythical-fred — APPROVED** [review]
Re-APPROVE on ea2ce7992a.

Since the prior review at af46f21d:
- Rebased onto 92f03be (EXCEPT/INTERSECT nullability fix merged into main). Rebase brings extra null-input filtering into `CalciteToDBSPCompiler` for n-ary Minus/Intersect, plus the corresponding `filterNonNullFields` signature refactor — orthogonal to this PR, already approved on its own merit.
- New `[SQL] Use stepWeightOne in more tests` commit: purely mechanical replacement of `step(... empty ...)` with `stepWeightOne(...)` across 8 test files. Empty-expected blocks shrink from `r | weight\n----` to `r\n---`. No behavioural change, no risk.
- New SetOpTests cases pick up my prior nit asking for a symmetric `testExceptAll*` ladder via the `checkSetOpNullTwoInts` helper (16-combination NULL cross-product for both EXCEPT and INTERSECT) plus targeted `testExceptMultiField`, `testExceptNullExcluded`, `testIntersectMultiField`. Good coverage.

Carry forward: my earlier soft note about `ExceptAllFoldRule` possibly being dead code (no MinusMergeRule in the pipeline today) stands but is non-blocking.

**Review by mythical-fred — APPROVED** [review]
Squashed to a single commit on top of the new base; the file-level diff lines up with what I already approved at `ea2ce79`. INTERSECT/EXCEPT ALL rewrite + the `SetOpTests` coverage from the prior round are intact. Carrying over the APPROVE — no new findings.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/optimizer/ExceptAllFoldRule.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        builder.minus(true);  // EXCEPT ALL
+
+        call.transformTo(builder.build());
+    }
+
+    public static final DefaultOptRuleConfig<ExceptAllFoldRule> CONFIG =
+            DefaultOptRuleConfig.<ExceptAllFoldRule>create()
+                    .withOperandSupplier(b ->
+                            b.operand(Minus.class)
+                             .predicate(m -> m.all && m.getInputs().size() > 2)
```
Comment (gz) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/calciteCompiler/optimizer/ExceptAllFoldRule.java:66]: this isnt ANSI Java 0.1 though

**Review by gz — APPROVED** [review]

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java [hunk 6]
```diff
@@ -1193,14 +1194,19 @@ private void visitMinus(LogicalMinus minus) {
...
+        DBSPSumOperator sum = new DBSPSumOperator(node, inputs);
+        final DBSPSimpleOperator result;
+        this.addOperator(sum);
         if (minus.all) {
-            throw new UnimplementedException("EXCEPT/MINUS ALL", 5483, node);
+            DBSPDifferentiateOperator diff = new DBSPDifferentiateOperator(node, sum.outputPort());
+            this.addOperator(diff);
+            DBSPPositiveOperator pos = new DBSPPositiveOperator(node, diff.outputPort());
+            this.addOperator(pos);
+            result = new DBSPIntegrateOperator(node.getFinal(), pos.outputPort());
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java:1205]: **Correctness concern for incremental updates.** The pattern here is `differentiate → positive → integrate` applied to `sum = A - B`.

In the CalciteToDBSP circuit (lifted level, full states), this computes:

```
I(positive(D(sum)))
```

which at time t equals `Σ_{i≤t} positive(sum[i] - sum[i-1])` — the accumulated positive *changes* to `A - B`, not `positive(A[t] - B[t])`.

For a single batch (t=0), `D(sum)[0] = sum[0]` and `I(…)[0] = positive(sum[0])`, so the result is correct. But for multi-step incremental updates with retractions, consider:

- t=0: A = {v:3}, B = {v:0} → sum = {v:3} → D = {v:3} → positive = {v:3} → I = {v:3} ✓
- t=1: A = {v:3}, B = {v:2} → sum = {v:1} → D = {v:-2} → positive = {v:0} → I = {v:3} ✗

Expected EXCEPT ALL at t=1: `positive({v:1})` = `{v:1}`, got `{v:3}`.

The D-I sandwich is the pattern for embedding an **incremental-only** operator (like `DBSPLeftJoinOperator`) in the lifted circuit. But `DBSPPositiveOperator` is a **lifted** operator (expanded by `DeltaExpandOperators`). The correct circuit for a lifted operator should be simply:

```java
result = new DBSPPositiveOperator(node, sum.outputPort());
```

without any `differentiate`/`integrate` wrapper — `DeltaExpandOperators` handles incrementalization via `integrate + binary_distinct(positive=true)`.

Alternatively, if the intent is to use positive as an incremental operator embedded in the lifted circuit, the sandwich direction should be `I → positive → D` (not `D → positive → I`).

All current tests pass because they are single-step (VALUES-based, append-only). Could you either:
1. Explain why `D → positive → I` is correct in the DBSP circuit model used here, or
2. Add a multi-step test with retractions (e.g., insert into A, then insert into B in a second step, verify the output retracts correctly)?

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java [hunk 7]
```diff
@@ -2429,40 +2435,65 @@ DBSPMapIndexOperator indexEntireTuple(CalciteRelNode node, OutputPort port, DBSP
...
+        this.addOperator(sum);
+
+        final DBSPSimpleOperator result;
+        if (all) {
+            DBSPDifferentiateOperator diff = new DBSPDifferentiateOperator(node, sum.outputPort());
+            this.addOperator(diff);
+            DBSPPositiveOperator pos = new DBSPPositiveOperator(node, diff.outputPort());
+            this.addOperator(pos);
+            result = new DBSPIntegrateOperator(node.getFinal(), pos.outputPort());
+        } else {
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java:2457]: Same `D → positive → I` pattern as in `visitMinus`. If the concern above is valid, this helper is also affected.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPPositiveOperator.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import org.dbsp.sqlCompiler.ir.expression.DBSPExpression;
+import org.dbsp.sqlCompiler.ir.type.DBSPType;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+public final class DBSPPositiveOperator extends DBSPUnaryOperator
+        implements IContainsIntegrator, IIncremental {
+    public DBSPPositiveOperator(CalciteRelNode node, OutputPort input) {
+        super(node, "positive", null, input.outputType(), false, input);
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPPositiveOperator.java:19]: Nit: the 5th constructor arg (`false`) sets `isMultiset = false`, meaning the output claims to be a set (weight ≤ 1 per element). But `positive` preserves original positive weights, which can be > 1. Should this be `input.isMultiset()` or just `true`?

This doesn't cause a build break because the `ToRustVisitor` handler omits `markDistinct`, but it could confuse downstream optimizations that inspect `isMultiset`.

File: sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/SetOpTests.java [hunk 9]
```diff
@@ -361,18 +410,113 @@ public void testIntersectMultiField() {
...
-                }
-            }
-        }
+    public void testIntersectAllBasic() {
+        // A = {1, 2, 2}  B = {2, 2, 3}
+        // Result = {2, 2}  (min(0,0)=0 for 1 and 3, min(2,2)=2 for 2)
+        this.qst("""
+                SELECT * FROM (VALUES (1), (2), (2)) AS T(x)
+                INTERSECT ALL
+                SELECT * FROM (VALUES (2), (2), (3)) AS T(x);
```
Comment (mythical-fred) [inline sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/SetOpTests.java:420]: Good test coverage for single-step semantics. Consider adding a multi-step incremental test that exercises retractions — e.g., insert data into one table, check the result, then insert into the other table and verify the output updates correctly. This would validate the `D → positive → I` pipeline under incremental updates, not just static VALUES.

**Review by mythical-fred — CHANGES_REQUESTED** [review]
**Correctness concern: the `D → positive → I` pattern needs justification for multi-step incremental updates.** See inline comments for details. The visitor plumbing is complete and `OptimizeDistinctVisitor` rules are sound, but I cannot approve until the core algorithm question is resolved.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java (hunk 6 above)
Comment (mihaibudiu) [reply sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java:1205]: Positive is treated exactly like Distinct everywhere, they are both-incremental only, as the interface IIncrementalOnly they inherit from shows.

**Review by mythical-fred — APPROVED** [review]
Retracting my correctness blocker — you're right and I was wrong.

I was reading `DBSPPositiveOperator` as a pointwise "map w -> max(0, w)" and then worrying that wrapping it with `D → · → I` computes the accumulated positive *changes* to `A - B`. But `positive` is an incremental-only operator, same as `distinct` (see `crates/dbsp/src/operator/dynamic/distinct.rs`: `PositiveWeight : DistinctSemantics` alongside `ZeroOrOneWeight`, both routed through `distinct_inner_generic` with an internal integral). The `D → op → I` sandwich is the standard total-from-incremental lifting; `distinct` is used the same way in this file and elsewhere, and gets it right. So `I(positive(D(sum(A, -B))))` is exactly `positive_total(A - B) = max(0, A - B)`, which is EXCEPT ALL. No bug.

Sorry for the noise. Approving.

The two smaller notes stand as non-blocking nits (the `isMultiset=false` question and the suggestion of a multi-step retraction test), but neither is a merge concern.

**Review by mythical-fred — APPROVED** [review]
Re-review at d55786eb (single squashed commit off same parent a7e4fd86 as c986a57 approval).

Delta since previously-approved c986a57 is small and clean:

1. **`CalciteToDBSPCompiler.except()`** gains a `needsDistinct` parameter. In `visitIntersect`'s rewrite `A EXCEPT (A EXCEPT B)`, the outer step now skips the distinct/positive lifting. This is sound:
   - INTERSECT ALL: `c_A − max(0, c_A − c_B) = min(c_A, c_B) ≥ 0`, so no clipping needed.
   - INTERSECT DISTINCT: inputs are pre-distinct (0/1 counts), so the raw sum is already the right zset.
2. **`DBSPTypeTuple.caster`**: fixes a latent bug where casting a nullable tuple to a non-nullable one previously called `to.none()` — invalid when `to.mayBeNull` is false. Now emits a `DBSPFailExpression("Cast to non-nullable value applied to NULL")`. Correct behavior.

Everything else (24 files) is unchanged from c986a57 and was reviewed there. Recap of the algorithm and plumbing:
- `EXCEPT ALL` = `I(positive(D(A + negate(B))))` — the `D → positive → I` sandwich is the standard total-from-incremental lifting; `positive` is `DistinctSemantics` in the DBSP crate, same interface as `distinct`.
- `INTERSECT ALL` = `A EXCEPT ALL (A EXCEPT ALL B)`.
- New `DBSPPositiveOperator` is fully plumbed through `AppendOnly`, `CircuitCloneVisitor`, `CircuitVisitor`, `KeyPropagation`, `Monotonicity`, `PropagateEmptySources`, `SeparateIntegrators`, `DeltaExpandOperators`, `InsertLimiters`, `OptimizeDistinctVisitor`, and `ToRustVisitor`.
- `DBSPBinaryDistinctOperator.positive` flag distinguishes Distinct- vs Positive-origin in the incremental expansion (used by `DeltaExpandOperators.postorder(DBSPPositiveOperator)` and threaded through `with(...)`/`fromJson`).
- `OptimizeDistinctVisitor` rules `distinct(positive)=distinct`, `positive(distinct)=distinct`, `positive(positive)=positive` are sound.
- `ExceptAllFoldRule` collapses n-ary EXCEPT ALL correctly.
- `SetOpTests` covers null / multi-column / three-way / min-counts / boundary cases.

CI: green. Docs (grammar / unsupported-operations) and Python negative tests (`test_grammar_tbl_fn.py`) are updated to reflect the new support. `MINUS ALL` remains a Calcite parse error (commented out with a note), which is acceptable — `EXCEPT ALL` works.

Ship it.
